### ISSUE TITLE
Generic/CyclomaticComplexity: add metrics

### ISSUE THREAD

**Issue by jrfnl** [PHPCSStandards/PHP_CodeSniffer#745]
## Is your feature request related to a problem?
PHPCS can provide statistics about a code base via the `--report=info` feature.

This information is recorded by sniffs via the `File::recordMetric()` method.

I believe it could be interesting for code base owners already using PHPCS to have a straight forward way to get insight into the cyclomatic complexity across their codebase.

## Describe the solution you'd like

For the sniff to start recording metrics, probably grouped in a few "steps", possibly along the lines of those mentioned in [this comment](https://github.com/squizlabs/PHP_CodeSniffer/issues/3501#issuecomment-994867608) (+ some more for very high complexity).

Maybe a grouping like this ?

* 1-5 (low risk)
* 6-10 (normal risk)
* 11-20 (moderate risk)
* 21-50 (high risk)
* 51-100 (very high risk)
* 101-500 (very high risk)
* 501-1000 (very high risk)
* 1001-5000 (very high risk)
* 5001-10000 (very high risk)
* 10000+ (very high risk)

- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md) and this is not a [support question](https://github.com/PHPCSStandards/PHP_CodeSniffer/discussions).
- [ ] I intend to create a pull request to implement this feature.

**Comment by paulshryock** [comment]
Adding a couple related resources, in case this is useful:

> Generally, 1-4 is low complexity, 5-7 indicates moderate complexity, 8-10 is high complexity, and 11+ is very high complexity.
> -- from https://phpmd.org/rules/codesize.html#cyclomaticcomplexity

> Limit the number of branch points per unit to 4.
> -- from "Building Maintainable Software" by Joost Visser, chapter 3: "Write simple units of code"
> https://ia801209.us.archive.org/12/items/BuildingMaintainableSoftwareSIG/Building_Maintainable_Software_SIG.pdf

**Comment by jrfnl** [comment]
@paulshryock Just to be sure we are talking about the same thing - this issue is **not** about how the cyclomatic complexity is calculated. That follows the industry standard as much as possible. And what is the recommended maximum complexity and absolute maximum complexity is configurable.

This issue is about reporting the overal state of a codebase, i.e. how many functions are there with a certain cyclomatic complexity ?
Now, while a `1-4`/`5-7`/`8-10`/`11+` grouping might make sense, this is not very informative if 100% of the codebase is `11+`.
To get a feel for the scale of the problem, you want more insight into real numbers, not just "too high", which would be similar to just running the sniff with a normal report.

**Comment by paulshryock** [comment]
> Just to be sure we are talking about the same thing

Ah, I did not realize the groupings referred to number of functions.

**Comment by jrfnl** [comment]
@paulshryock I still have a feeling we're talking about different things.

The `--report=info` (metrics) feature shows statistics about the code base.

For example:
```
Line length:
        80 or less  => 88,660 ( 94.23%)
        81-120      =>  4,842 (  5.15%)
        121-150     =>    452 (  0.48%)
        151 or more =>    138 (  0.15%)
        --------------------------------
        total       => 94,092 (100.00%)
```

So, the metrics for cyclomatic complexity would show a grouping a how many functions in a codebase have which level of complexity. Something like:
```
Cyclomatic complexity:
        1-5 (low risk)            => 528 ( 50.14%)
        6-10 (normal risk)        => 432 ( 41.03%)
        11-20 (moderate risk)     =>  74 (  7.03%)
        21-50 (high risk)         =>  18 (  1.71%)
        501-1000 (very high risk) =>   1 (  0.09%)
        --------------------------------
        total                     => 1.053 (100.00%)
```

Does that help ?

**Comment by paulshryock** [comment]
> the metrics for cyclomatic complexity would show a grouping a how many functions

Yeah, I see what you mean. When I left my first comment, I didn't realize the issue was about number of functions. I thought it was about number of branch points in a single function, as I thought was referenced in [this linked comment](https://github.com/squizlabs/PHP_CodeSniffer/issues/3501#issuecomment-994867608). Sorry for the noise, please disregard my comment or mark off-topic.
